### PR TITLE
lib/tflm: Skip batch size when reading input/output shapes.

### PR DESCRIPTION
**Pull request by iabdalkader**

### CODE REVIEW

**Comment by kwagyeman** [comment]
MicroSpeech doesn't work because the audio-preprocessing module doesn't have a batch size...

input->dims->size: 2
input->dims->data[0]: 1
input->dims->data[1]: 480
output->dims->size: 1
output->dims->data[0]: 40

Will suggest changes.

File: src/lib/tflm/tflm_backend.cc [hunk 1]
```diff
@@ -227,8 +227,13 @@ int ml_backend_init_model(py_ml_model_obj_t *model) {
...
 
-        mp_obj_tuple_t *o = (mp_obj_tuple_t *) MP_OBJ_TO_PTR(mp_obj_new_tuple(input->dims->size, NULL));
-        for (int j=0; j<input->dims->size; j++) {
+        // Check input data shape.
+        if (input->dims->size < 2 || input->dims->data[0] != 1) {
+            mp_raise_msg(&mp_type_ValueError, MP_ERROR_TEXT("Unexpected input shape"));
+        }
+
+        mp_obj_tuple_t *o = (mp_obj_tuple_t *) MP_OBJ_TO_PTR(mp_obj_new_tuple(input->dims->size - 1, NULL));
+        for (int j=1; j<input->dims->size; j++) {
```
Comment (kwagyeman) [inline src/lib/tflm/tflm_backend.cc:236]:
```suggestion
        // Check input data shape.
        int skip_batch = input->dims->size > 1;
        if (skip_batch && input->dims->data[0] != 1) {
            mp_raise_msg(&mp_type_ValueError, MP_ERROR_TEXT("Batch size must be 1"));
        }

        mp_obj_tuple_t *o = (mp_obj_tuple_t *) MP_OBJ_TO_PTR(mp_obj_new_tuple(input->dims->size - skip_batch, NULL));
        for (int j=skip_batch; j<input->dims->size; j++) {
```

File: src/lib/tflm/tflm_backend.cc [hunk 2]
```diff
@@ -253,8 +258,13 @@ int ml_backend_init_model(py_ml_model_obj_t *model) {
...
 
-        mp_obj_tuple_t *o = (mp_obj_tuple_t *) MP_OBJ_TO_PTR(mp_obj_new_tuple(output->dims->size, NULL));
-        for (int j=0; j<output->dims->size; j++) {
+        // Check output data shape.
+        if (output->dims->size < 2 || output->dims->data[0] != 1) {
+            mp_raise_msg(&mp_type_ValueError, MP_ERROR_TEXT("Unexpected output shape"));
+        }
+
+        mp_obj_tuple_t *o = (mp_obj_tuple_t *) MP_OBJ_TO_PTR(mp_obj_new_tuple(output->dims->size - 1, NULL));
+        for (int j=1; j<output->dims->size; j++) {
```
Comment (kwagyeman) [inline src/lib/tflm/tflm_backend.cc:267]:
```suggestion
        // Check output data shape.
        int skip_batch = output->dims->size > 1;
        if (skip_batch && output->dims->data[0] != 1) {
            mp_raise_msg(&mp_type_ValueError, MP_ERROR_TEXT("Unexpected output shape"));
        }

        mp_obj_tuple_t *o = (mp_obj_tuple_t *) MP_OBJ_TO_PTR(mp_obj_new_tuple(output->dims->size - skip_batch, NULL));
        for (int j=skip_batch; j<output->dims->size; j++) {
```

File: src/lib/tflm/tflm_backend.cc [hunk 3]
```diff
@@ -227,8 +227,13 @@ int ml_backend_init_model(py_ml_model_obj_t *model) {
...
-        mp_obj_tuple_t *o = (mp_obj_tuple_t *) MP_OBJ_TO_PTR(mp_obj_new_tuple(input->dims->size, NULL));
-        for (int j=0; j<input->dims->size; j++) {
+        // Check input data shape.
+        if (input->dims->size < 2 || input->dims->data[0] != 1) {
+            mp_raise_msg(&mp_type_ValueError, MP_ERROR_TEXT("Unexpected input shape"));
+        }
+
+        mp_obj_tuple_t *o = (mp_obj_tuple_t *) MP_OBJ_TO_PTR(mp_obj_new_tuple(input->dims->size - 1, NULL));
+        for (int j=1; j<input->dims->size; j++) {
             o->items[j] = mp_obj_new_int(input->dims->data[j]);
```
Comment (kwagyeman) [inline src/lib/tflm/tflm_backend.cc:237]:
```suggestion
        // Check input data shape.
        int skip_batch = input->dims->size > 1;
        if (skip_batch && input->dims->data[0] != 1) {
            mp_raise_msg(&mp_type_ValueError, MP_ERROR_TEXT("Batch size must be 1"));
        }

        mp_obj_tuple_t *o = (mp_obj_tuple_t *) MP_OBJ_TO_PTR(mp_obj_new_tuple(input->dims->size - skip_batch, NULL));
        for (int j=0; j<input->dims->size-skip_batch; j++) {
            o->items[j] = mp_obj_new_int(input->dims->data[j + skip_batch]);
        }
```

File: src/lib/tflm/tflm_backend.cc [hunk 4]
```diff
@@ -253,8 +258,13 @@ int ml_backend_init_model(py_ml_model_obj_t *model) {
...
-        mp_obj_tuple_t *o = (mp_obj_tuple_t *) MP_OBJ_TO_PTR(mp_obj_new_tuple(output->dims->size, NULL));
-        for (int j=0; j<output->dims->size; j++) {
+        // Check output data shape.
+        if (output->dims->size < 2 || output->dims->data[0] != 1) {
+            mp_raise_msg(&mp_type_ValueError, MP_ERROR_TEXT("Unexpected output shape"));
+        }
+
+        mp_obj_tuple_t *o = (mp_obj_tuple_t *) MP_OBJ_TO_PTR(mp_obj_new_tuple(output->dims->size - 1, NULL));
+        for (int j=1; j<output->dims->size; j++) {
             o->items[j] = mp_obj_new_int(output->dims->data[j]);
```
Comment (kwagyeman) [inline src/lib/tflm/tflm_backend.cc:268]:
```suggestion
        // Check output data shape.
        int skip_batch = output->dims->size > 1;
        if (skip_batch && output->dims->data[0] != 1) {
            mp_raise_msg(&mp_type_ValueError, MP_ERROR_TEXT("Batch size must be 1"));
        }

        mp_obj_tuple_t *o = (mp_obj_tuple_t *) MP_OBJ_TO_PTR(mp_obj_new_tuple(output->dims->size - skip_batch, NULL));
        for (int j=0; j<output->dims->size-skip_batch; j++) {
            o->items[j] = mp_obj_new_int(output->dims->data[j + skip_batch]);
        }
```

**Comment by kwagyeman** [comment]
Microspeech works with changes.

**Comment by iabdalkader** [comment]
> MicroSpeech doesn't work because the audio-preprocessing module has no output batch size...

It definitely does:
```
{size: 7760, ram: 8992, input_shape: ((1, 480),), input_scale: (1.0,), input_zero_point: (0,), input_dtype: ('h',), output_shape: ((40,),), output_scale: (1.0,), output_zero_point: (0,), output_dtype: ('f',)}
```

I messed up the output. The output is the one that never has any batch size :)

**Comment by iabdalkader** [comment]
Fixed, can you test again please just to be sure ? @kwagyeman

**Comment by iabdalkader** [comment]
That was wrong too sorry for the confusion. I need to think about this some more.

**Comment by kwagyeman** [comment]
Okay, let me know when to test.

File: src/lib/tflm/tflm_backend.cc (hunk 3 above)
Comment (kwagyeman) [inline src/lib/tflm/tflm_backend.cc:237]:
```suggestion
        for (int j=1; j<input->dims->size; j++) {
            o->items[j-1] = mp_obj_new_int(input->dims->data[j]);
```

**Comment by kwagyeman** [comment]
@iabdalkader Yeah, I was thinking about this more and we should just increase the size of the max dims in the ND array. Doing this is the best fix.

**Comment by iabdalkader** [comment]
closing then.

**Pull request closed without merging**